### ISSUE TITLE
`send_datagram_chunks`

### ISSUE THREAD

**Issue by kixelated** [quinn-rs/quinn#1724]
I'm adding datagram support to `webtransport-quinn`. WebTransport uses a VarInt in front of each datagram to identify the session. My plan is to mirror the quinn API, something like:
```rust
pub async fn send_datagram(&self, data: Bytes) -> Result<(), SessionError> {
```

I want the `session_id` prefix to be transparent to the application. Unfortunately, with this API it's impossible to avoid making a copy of the data since the current Quinn `send_datagram` method requires the payload to be continuous in memory.

Proposal: a `send_datagram_chunks` method that mimicks the `SendStream` API.
```rust
pub async fn send_datagram_chunks(
    &mut self,
    bufs: &mut [Bytes]
) -> Result<(), SendDatagramError>
```

**Comment by Ralith** [comment]
I'm not sure allocating a separate `Bytes` for a `VarInt` tag will actually improve performance vs. copying, since datagrams have to be fairly small to begin with. Do you have any data?

**Comment by kixelated** [comment]
> I'm not sure allocating a separate `Bytes` for a `VarInt` tag will actually improve performance vs. copying, since datagrams have to be fairly small to begin with. Do you have any data?

Sorry, I updated the description with some more information and less of a proposal.

The issue is that I need to allocate a copy of each datagram since there's no way to prepend/append otherwise. I don't have any performance numbers but this would undoubtedly be an issue at high throughput.

**Comment by djc** [comment]
I think this is already supported with the current API? A `Bytes` [is](https://docs.rs/bytes/latest/bytes/struct.Bytes.html):

> A cheaply cloneable and sliceable chunk of contiguous memory.

That is, it should be possible to allocate a single `Bytes` for holding all your datagrams, than split it up and pass each part (which is really a view into the same underlying allocation) into `Datagrams::send()`.

**Comment by kixelated** [comment]
> I think this is already supported with the current API? A `Bytes` [is](https://docs.rs/bytes/latest/bytes/struct.Bytes.html):
> 
> > A cheaply cloneable and sliceable chunk of contiguous memory.
> 
> That is, it should be possible to allocate a single `Bytes` for holding all your datagrams, then split it up and pass each part (which is really a view into the same underlying allocation) into `Datagrams::send()`.

Not quite the same issue. The key thing is that I'm writing a library, so I don't actually create the datagram payload. I want my library to add the WebTransport header to each datagram.

You're right that the application could encode the `session_id` header into each datagram, but I really don't want the application to know anything about WebTransport internals. It's meant to be transparent to the application so the WebTransport API is identical to QUIC.

A similar issue is fragmentation. Let's say I have a 1MB video `Bytes` blob that I want to fragment into datagrams. You're right that slicing is free, however that doesn't work since each datagram needs some sort of header (ex. sequence number), otherwise reassembly is impossible. You run into the same problem and need to allocate a copy of each datagram on the heap.

Meanwhile, the streams API lets you avoid this. You can do two separate writes: one for the header and one for the `Bytes` payload.

**Comment by Ralith** [comment]
Yeah, that generally makes sense. I'm curious how much the extra API complexity would actually buy us, though. Streams can have much larger individual writes, though I guess in either case you might be sending up to at most the link capacity, so maybe it's not so different...

I'm open to having this.

**Comment by kixelated** [comment]
Yeah I don't think there's any urgency for this API. I'm not using datagrams and actively advocate against them.

But this would be a blocker for using datagrams at scale. Allocating on the heap for every packet gets expensive. I'm just going to leave a comment in my library and link to this issue.

**Comment by cameronelliott** [comment]
Just dropping by, but to do this kind of thing right, IMHO, you really want the consumers of the API to do ALL packet allocation for both RX & TX packets, and present the buffers to packages like Quinn. I understand this may be radically different than the existing API. 
The big benefit is that, if done correctly (big if) consumers of Quinn can then use facilities like AF_XDP and DPDK without ever copying packets. Ping me if you want more info.
